url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/fsa/tffst/Tffst.java

```diff
@@ -0,0 +1,1712 @@
+//    while (i.hasNext()) {
+//      // System.out.println(number);
+//      State s = i.next();
+//      s.setNumber(number++);
+//    }
+//  }
+
+  /**
+   * Returns <a href="http://www.research.att.com/sw/tools/graphviz/"
+   * target="_top">Graphviz Dot </a> representation of this Tffst.
+   */
+  public String toDot(String label) {
+     return toDot(label, "TB"); 
+  }
+
+  
+  
+  /**
+   * Returns <a href="http://www.research.att.com/sw/tools/graphviz/"
+   * target="_top">Graphviz Dot </a> representation of this Tffst.
+   */
+  public String toDot(String label, String orientation) {
+
+    StringBuffer b = new StringBuffer("digraph Tffst {\n"); 
+    
+    b.append("graph [ bgcolor=white,fontname=Arial, fontcolor=black, fontsize=12 , "
+    	+ "overlap=compress, ranksep=\"0.3\", nodesep=\"0.1\", splines=true];\n");
+    b.append("node [ fontname=Arial, fontcolor=black, fontsize=10];\n");
+    b.append("edge [ fontname=Helvetica, fontcolor=black, fontsize=11, arrowsize=0.6, color=grey ];\n");
+    
+    if (orientation.equalsIgnoreCase("LR")) b.append("rankdir = \"LR\";\n");
+    
+    b.append("label = \"").append(label).append("\";\n");
+    
+    Iterator<State> i = getStates().iterator();
+    while (i.hasNext()) {
+      State s = i.next();
+      b.append("  ").append(s.getNumber()); 
+      if (s.accept) 
+        b.append(" [shape=doublecircle,label=\"").append(s.getLabel()).append("\"];\n"); 
+//     
+//      if (s == initial)
+//        b.append(" [shape=circle, style=filled, fillcolor=grey, abel=\"").append(s.getLabel()).append("\"];\n"); 
+//      
+      else
+        b.append(" [shape=circle,label=\"").append(s.getLabel()).append("\"];\n"); 
+      if (s == initial) {
+        b.append("  initial [shape=plaintext,label=\"\"];\n"); 
+        b.append("  initial -> ").append(s.getNumber()).append("\n"); 
+      }
+      Iterator<Transition> j = s.getTransitionsIterator();
+      while (j.hasNext()) {
+        Transition t = j.next();
+        b.append("  ").append(s.getNumber()); 
+        t.appendDot(b);
+      }
+    }
+    return b.append("}\n").toString(); 
+  }
+
+  /**
+   * returns a Tffst representing this tffst but with simple input and output labels on each edge. 
+   * it expands TfStrings in simple pairs of TFs adding epsilons
+   * 
+   * @return
+   */
+  public Tffst toSingleLabelTransitions() {
+	
+	Tffst simpleTffst = new Tffst();
+	HashMap<State, State> m = new HashMap<State, State>();
+
+	State last;
+	Set<State> states = getStates();
+
+	for (State s : states) {
+
+	  if (m.get(s) == null) {
+		last = new State();
+		m.put(s, last);
+		last.accept = s.accept;
+		if (s == initial)
+		  simpleTffst.initial = last;
+	  } else 
+		last = m.get(s);
+
+	  for (Transition t : s.getTransitions()) {
+
+		if (m.get(t.getTo()) == null) {
+		  State newTo = new State();
+		  m.put(t.getTo(), newTo);
+		  newTo.accept = t.getTo().accept;
+		  if (t.getTo() == initial)
+			  simpleTffst.initial = t.getTo();
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1109:1115, 1178:1184.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.